src/System.Linq.Expressions/src/System/Linq/Expressions/BlockExpression.cs

```diff
@@ -88,9 +88,37 @@ public override Type Type
         /// <returns>This expression if no children changed, or an expression with the updated children.</returns>
         public BlockExpression Update(IEnumerable<ParameterExpression> variables, IEnumerable<Expression> expressions)
         {
-            if (variables == Variables && expressions == Expressions)
+            bool variablesMatch;
+            if (variables == Variables)
+                variablesMatch = true;
+            else if (Variables == null || Variables.Count == 0)
+                variablesMatch = variables == null || !variables.Any();
+            else if (variables == null)
+                variablesMatch = false;
+            else
             {
-                return this;
+                // Just in case variables cannot be streams and cannot be repeated.
+                IList<ParameterExpression> varList = variables as IList<ParameterExpression> ?? variables.ToReadOnly();
+                variablesMatch = Variables.Count == varList.Count && Variables.SequenceEqual(varList);
+            }
+
+            if(variablesMatch)
+            {
+                bool expressionsMatch;
+                if (expressions == Expressions)
+                    expressionsMatch = true;
+                else if (expressions == null)
```

Comment:
Even worse, I jump through all these hoops to short circuit with as little allocation as possible, and then the final call to Expression.Block is incorrect.